x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -853,6 +856,69 @@ protected String parseSourceValue(Object value) {
             };
         }
 
+        @Override
+        public QueryableExpression asQueryableExpression(SearchExecutionContext context) {
+            if (false == isSearchable()) {
+                throw new IllegalStateException("Wildcard fields are always searchable.");
+            }
+            return StringQueryableExpression.field(name(), new StringQueryableExpression.Queries() {
+                @Override
+                public Query approximateExists() {
+                    return existsQuery(context);
+                }
+
+                @Override
+                public Query approximateTermQuery(String term) {
+                    // A term query without any rechecking. The script will recheck.
+                    Set<String> ngrams = new HashSet<>();
+                    getNgramTokens(ngrams, addLineEndChars(term));
+
+                    BooleanQuery.Builder approximation = new BooleanQuery.Builder();
+                    int clauseCount = 0;
+                    for (String string : ngrams) {
+                        if (clauseCount >= MAX_CLAUSES_IN_APPROXIMATION_QUERY) {
+                            break;
+                        }
+                        addClause(string, approximation, Occur.MUST);
+                        clauseCount++;
+                    }
+                    if (clauseCount == 0) {
+                        return new MatchAllDocsQuery();
+                    }
+
+                    return approximation.build();
+                }
+
+                @Override
+                public Query approximateSubstringQuery(String term) {
+                    if (false == UnicodeUtil.validUTF16String(term)) {
+                        return new MatchAllDocsQuery();
+                    }
+                    Set<String> ngrams = new HashSet<>();
+                    getNgramTokens(ngrams, term);
+
+                    BooleanQuery.Builder approximation = new BooleanQuery.Builder();
+                    int clauseCount = 0;
+                    for (String string : ngrams) {
+                        if (clauseCount >= MAX_CLAUSES_IN_APPROXIMATION_QUERY) {
+                            break;
+                        }
+                        if (string.length() < NGRAM_SIZE) {
+                            continue;
+                        }
+                        // We're not using addClause here because the
```

Comment:
Hmmm - looks like I lost my train of thought. I believe this is because only even want the term queries. I think we've filtered out all the cases that'll cause other things, but this feels a little more readable this way.

Reply:
addClause will expand a search for `ab` as `ab*` prefix query on the ngram index because we only index 3grams in the approximation index. Not sure if that's an optimisation you want to keep otherwise 2-character queries won't be accelerated at all? We took the decision not to try accelerate single-character searches just because that could be a lot of ngram matching for little overall acceleration.

Reply:
I think I was worried about the start and end markers. I'm not against uses `ab*` in theory, though it was nice that queries targeting runtime fields backed by wildcards would never have to be rewritten. OTOH, we aren't likely to have that many trigrams matching any particular bigram.